

## Assignment:  Halloween Candy

### Name: Armonti Du-Bose-Hill

## Overview

Data cleaning and data wrangling from a survey of Halloween candy to prepare it for a machine learning project.

## Data Set

The data set that we will be using is the 2017 Halloween Candy Hierarchy data set as discussed in this [boingboing](https://boingboing.net/2017/10/30/the-2017-halloween-candy-hiera.html) article.  You can also read more about the data in the [Science Creative Quarterly](https://www.scq.ubc.ca/so-much-candy-data-seriously/).

The following are the rating instructions from the survey:  

> Basically, consider that feeling you get when you receive this item in your Halloween haul. Does it make you really happy (JOY)? Or is it something that you automatically place in the junk pile (DESPAIR)? MEH for indifference, and you can leave blank if you have no idea what the item is.

This data is a great example of a messy data set, especially since they allowed respondents to enter text for a number of the fields. Also, note that some of the comments in the file might be considered inappropriate to some readers but cleaning this type of data is normal in a lot of data science projects.

## Note

<u>Show Work</u>

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22
```

## Our End Goal

Our end goal for this project is to clean the data so that we could then create a machine learning model. We want to see if we are able to predict a person's gender based purely on their candy preferences.

## Initial Import & Exploration

In [1]:
# initial imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', 20)
import warnings
warnings.filterwarnings("ignore")

***Import the data and create a DataFrame called `candy`.  We need to include `encoding='iso-8859-1'` during the import because there are special characters in the data that Pandas doesn't recognize. This happens a lot when attempting to import data where the public is able to input answers, especially if there are foreign language characters included. The normal encoding for Pandas is `utf-8`, so changing the encoding allows Pandas to recognize those special characters.

In [2]:
# read_csv with iso-8859-1 encoding; using latin-1 would also work here
candy_full = pd.read_csv('candy.csv', encoding='iso-8859-1')

# copy to new DF so that we can have a copy of the original import if needed
candy = candy_full.copy()

Take a brief look at the data.

In [3]:
# first five rows
candy.head()

Internal ID Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY  \
0     90258773            NaN        NaN     NaN         NaN   
1     90272821             No       Male      44        USA    
2     90272829            NaN       Male      49         USA   
3     90272840             No       Male      40          us   
4     90272841             No       Male      23         usa   

  Q5: STATE, PROVINCE, COUNTY, ETC Q6 | 100 Grand Bar  \
0                              NaN                NaN   
1                               NM                MEH   
2                         Virginia                NaN   
3                               or                MEH   
4                         exton pa                JOY   

  Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                NaN                                       
1                                            DESPAIR                                       
2                                                NaN                                       
3                                            DESPAIR                                       
4                                            DESPAIR                                       

  Q6 | Any full-sized candy bar Q6 | Black Jacks  ... Q8: DESPAIR OTHER  \
0                           NaN              NaN  ...               NaN   
1                           JOY              MEH  ...               NaN   
2                           NaN              NaN  ...               NaN   
3                           JOY              MEH  ...               NaN   
4                           JOY          DESPAIR  ...               NaN   

                                  Q9: OTHER COMMENTS      Q10: DRESS  \
0                                                NaN             NaN   
1  Bottom line is Twix is really the only candy w...  White and gold   
2                                                NaN             NaN   
3                             Raisins can go to hell  White and gold   
4                                                NaN  White and gold   

  Unnamed: 113 Q11: DAY Q12: MEDIA [Daily Dish] Q12: MEDIA [Science]  \
0          NaN      NaN                     NaN                  NaN   
1          NaN   Sunday                     NaN                  1.0   
2          NaN      NaN                     NaN                  NaN   
3          NaN   Sunday                     NaN                  1.0   
4          NaN   Friday                     NaN                  1.0   

  Q12: MEDIA [ESPN] Q12: MEDIA [Yahoo] Click Coordinates (x, y)  
0               NaN                NaN                      NaN  
1               NaN                NaN                 (84, 25)  
2               NaN                NaN                      NaN  
3               NaN                NaN                 (75, 23)  
4               NaN                NaN                 (70, 10)  

[5 rows x 120 columns]

Run a code to see information about the DataFrame.



In [4]:
# check info about the DataFrame
candy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2479 entries, 0 to 2478
Columns: 120 entries, Internal ID to Click Coordinates (x, y)
dtypes: float64(4), int64(1), object(115)
memory usage: 2.3+ MB


According to the Pandas documentation:  "If the DataFrame has more than max_cols columns, the truncated output is used. By default, the setting in pandas.options.display.max_info_columns is used."

Make the columns display by setting the `max_cols` argument equal to the number of columns in the data set.

In [5]:
# check info, set max_cols
candy.info(max_cols=120)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2479 entries, 0 to 2478
Data columns (total 120 columns):
 #    Column                                                                                 Non-Null Count  Dtype  
---   ------                                                                                 --------------  -----  
 0    Internal ID                                                                            2479 non-null   int64  
 1    Q1: GOING OUT?                                                                         2368 non-null   object 
 2    Q2: GENDER                                                                             2437 non-null   object 
 3    Q3: AGE                                                                                2394 non-null   object 
 4    Q4: COUNTRY                                                                            2414 non-null   object 
 5    Q5: STATE, PROVINCE, COUNTY, ETC                                   

Of course, if you are just looking for the column names, you can just use a simple `for` loop.

In [6]:
# print a list of column names
for col in candy.columns:
    print(col)

Internal ID
Q1: GOING OUT?
Q2: GENDER
Q3: AGE
Q4: COUNTRY
Q5: STATE, PROVINCE, COUNTY, ETC
Q6 | 100 Grand Bar
Q6 | Anonymous brown globs that come in black and orange wrappers	(a.k.a. Mary Janes)
Q6 | Any full-sized candy bar
Q6 | Black Jacks
Q6 | Bonkers (the candy)
Q6 | Bonkers (the board game)
Q6 | Bottle Caps
Q6 | Box'o'Raisins
Q6 | Broken glow stick
Q6 | Butterfinger
Q6 | Cadbury Creme Eggs
Q6 | Candy Corn
Q6 | Candy that is clearly just the stuff given out for free at restaurants
Q6 | Caramellos
Q6 | Cash, or other forms of legal tender
Q6 | Chardonnay
Q6 | Chick-o-Sticks (we donÕt know what that is)
Q6 | Chiclets
Q6 | Coffee Crisp
Q6 | Creepy Religious comics/Chick Tracts
Q6 | Dental paraphenalia
Q6 | Dots
Q6 | Dove Bars
Q6 | Fuzzy Peaches
Q6 | Generic Brand Acetaminophen
Q6 | Glow sticks
Q6 | Goo Goo Clusters
Q6 | Good N' Plenty
Q6 | Gum from baseball cards
Q6 | Gummy Bears straight up
Q6 | Hard Candy
Q6 | Healthy Fruit
Q6 | Heath Bar
Q6 | Hershey's Dark Chocolate
Q6 | HersheyÕ

This data set is pretty messy. Your goal is now to perform the following actions to get it to the point where it can be passed to a machine learning model.

<span style='color:red'>**Note: Unless the instructions ask you to do something different, please always update the original `candy` DataFrame for the exercises below.

## Data Cleaning

**Exercise1:** Taking a look at the column names, you may notice that some include the character `Õ`. This should instead be an apostrophe `'` mark. Rename the column names that include the `Õ` character and replace it with an apostrophe.  

Remember that you should be updating the `candy` DataFrame for the tasks listed as unless told differently.

In [7]:
candy.columns = candy.columns.str.replace("Õ", "'", regex = False)

**Q1:** How many duplicated rows are there in the file? Assume that a duplicate is any row that is *exactly* the same as another one. Save this number as `Q1`.

In [8]:
Q1 = candy.duplicated().sum()

**Q2:** How many duplicated rows are there in the file if we were to assume that a duplicate is any row with the same `Internal ID` number as another. In other words, even if the other values are different, a row would count as a duplicate if it had the same `Internal ID` as another. Save this number as `Q2`.

In [9]:
Q2 = candy['Internal ID'].duplicated().sum()

**Exercise2:** Drop any duplicates from the `candy` DataFrame.  Duplicates are to be defined as any row with the same `Internal ID` as another. Use the default setting that keeps the first record from the duplicates.

In [10]:
candy = candy.drop_duplicates('Internal ID', keep= 'first')

**Exercise3:** Your next task is to remove the following columns from the `candy` DataFrame as we will not use these columns for this project.  You are welcome to do further analysis on these columns but do not save your analysis in this notebook.

Remove the following columns: `Internal ID`, `Q5: STATE, PROVINCE, COUNTY, ETC`, `Q7: JOY OTHER`, `Q8: DESPAIR OTHER`, `Q9: OTHER COMMENTS`, `Unnamed: 113`, `Click Coordinates (x, y)`.

In [11]:
columns_removed = [
    'Internal ID',
     'Q5: STATE, PROVINCE, COUNTY, ETC',
     'Q7: JOY OTHER',
     'Q8: DESPAIR OTHER',
     'Q9: OTHER COMMENTS',
     'Unnamed: 113',
     'Click Coordinates (x, y)'
]

candy = candy.drop(columns= columns_removed)
candy

Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY Q6 | 100 Grand Bar  \
0               NaN        NaN     NaN         NaN                NaN   
1                No       Male      44        USA                 MEH   
2               NaN       Male      49         USA                NaN   
3                No       Male      40          us                MEH   
4                No       Male      23         usa                JOY   
...             ...        ...     ...         ...                ...   
2474             No       Male      24         USA                JOY   
2475             No     Female      33         USA                MEH   
2476             No     Female      26        USA                 MEH   
2477             No       Male      58         Usa                NaN   
2478             No     Female      66         usa            DESPAIR   

     Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
0                                                   NaN                                       
1                                               DESPAIR                                       
2                                                   NaN                                       
3                                               DESPAIR                                       
4                                               DESPAIR                                       
...                                                 ...                                       
2474                                            DESPAIR                                       
2475                                            DESPAIR                                       
2476                                            DESPAIR                                       
2477                                                NaN                                       
2478                                            DESPAIR                                       

     Q6 | Any full-sized candy bar Q6 | Black Jacks Q6 | Bonkers (the candy)  \
0                              NaN              NaN                      NaN   
1                              JOY              MEH                  DESPAIR   
2                              NaN              NaN                      NaN   
3                              JOY              MEH                      MEH   
4                              JOY          DESPAIR                      MEH   
...                            ...              ...                      ...   
2474                           MEH          DESPAIR                  DESPAIR   
2475                           JOY              NaN                      NaN   
2476                           JOY          DESPAIR                      MEH   
2477                           NaN              NaN                      NaN   
2478                           JOY          DESPAIR                  DESPAIR   

     Q6 | Bonkers (the board game)  ... Q6 | Whatchamacallit Bars  \
0                              NaN  ...                       NaN   
1                          DESPAIR  ...                   DESPAIR   
2                              NaN  ...                       NaN   
3                          DESPAIR  ...                       JOY   
4                          DESPAIR  ...                       JOY   
...                            ...  ...                       ...   
2474                           MEH  ...                   DESPAIR   
2475                           NaN  ...                       JOY   
2476                           JOY  ...                       MEH   
2477                           NaN  ...                       NaN   
2478                       DESPAIR  ...                   DESPAIR   

     Q6 | White Bread Q6 | Whole Wheat anything Q6 | York Peppermint Patties  \
0                 NaN                       NaN                          NaN   
1             DESPAIR                

**Code Check:** As a check for the above exercises, the shape of your data should now be: `(2460, 113)`

In [12]:
candy.shape

(2460, 113)

**Exercise4:** Let's now take a look at the `Q2: GENDER` column since this will be what we are trying to predict. Take a look at the value counts for this column.

In [13]:
candy['Q2: GENDER'].value_counts()

Q2: GENDER
Male                  1466
Female                 839
I'd rather not say      83
Other                   30
Name: count, dtype: int64

**Q3:** How many missing values are in the `Q2: GENDER` column? Save this as `Q3`.

In [14]:
Q3 = candy['Q2: GENDER'].isna().sum()

**Exercise5:** Using the `candy` DataFrame, remove all rows with a missing value in the `Q2: GENDER` column.  (This should overwrite and be saved as `candy` like you have been doing for the previous exercises.)

In [15]:
candy = candy.dropna(subset= ['Q2: GENDER'])

**Exercise6:** For this project, we want to use binary classification, which predicts one of two classes. Because we are using a binary classification model, we will try to predict between `Male` or `Female`. Select only the rows that contain either `Male` or `Female` in the `Q2: GENDER` column.

In [16]:
candy = candy[candy['Q2: GENDER'].isin(['Male', 'Female'])]

**Code Check:** As a check for the above exercises, the shape of your data should now be: `(2305, 113)`

In [17]:
candy.shape

(2305, 113)

Now, let's work on filling some of the missing data.  There are easier ways to do this with the sklearn library which you will learn about more in the machine learning classes, but for now, let's try to practice our Pandas skills.

**Q4:** How many missing values are in the `Q1: GOING OUT?` column? Save this number as `Q4`.

In [18]:
Q4 = candy['Q1: GOING OUT?'].isna().sum()

**Exercise7:** For a future analysis question, we are interested in those that we know will *definitely* go out for Halloween.  Because of this, fill all missing values in the `Q1: GOING OUT?` column with a `No` value.

In [19]:
candy['Q1: GOING OUT?'].fillna('No', inplace= True)

**Code Check:** Double check your above work and look at the value counts for the `Q1: GOING OUT?` column.  Make sure that you only have "Yes" and No" values and that they add up to 2305, which is the number of rows you should have at this step in the assignment.

In [20]:
candy['Q1: GOING OUT?'].value_counts()


Q1: GOING OUT?
No     2007
Yes     298
Name: count, dtype: int64

**Exercise8:** To get ready for the next step, let's practice selecting all the columns: going from `Q6 | 100 Grand Bar` to `Q11: DAY`.  Save this slice as `candy_slice`.

In [21]:
candy_slice = candy.loc[:, 'Q6 | 100 Grand Bar' : 'Q11: DAY']


**Exercise9:** Now that you know how to slice the data, fill any missing values in the `candy` DataFrame for those columns (going from `Q6 | 100 Grand Bar` to `Q11: DAY`) with the string `NO_ANSWER`. **Make sure you are working with the `candy` DataFrame and not the `candy_slice` DataFrame.**

In [22]:
candy.loc[:, 'Q6 | 100 Grand Bar' : 'Q11: DAY'] = candy.loc[:, 'Q6 | 100 Grand Bar' : 'Q11: DAY'].fillna('NO_ANSWER')

**Exercise10:** For all four `Q12: Media` columns in the `candy` DataFrame, fill the missing values with `0.0`.

In [23]:
candy.loc[:, candy.columns.str.startswith('Q12: MEDIA')] = candy.loc[:, candy.columns.str.startswith('Q12: MEDIA')].fillna(0.0)
candy

Q1: GOING OUT? Q2: GENDER Q3: AGE Q4: COUNTRY Q6 | 100 Grand Bar  \
1                No       Male      44        USA                 MEH   
2                No       Male      49         USA          NO_ANSWER   
3                No       Male      40          us                MEH   
4                No       Male      23         usa                JOY   
5                No       Male     NaN         NaN                JOY   
...             ...        ...     ...         ...                ...   
2474             No       Male      24         USA                JOY   
2475             No     Female      33         USA                MEH   
2476             No     Female      26        USA                 MEH   
2477             No       Male      58         Usa          NO_ANSWER   
2478             No     Female      66         usa            DESPAIR   

     Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)  \
1                                               DESPAIR                                       
2                                             NO_ANSWER                                       
3                                               DESPAIR                                       
4                                               DESPAIR                                       
5                                               DESPAIR                                       
...                                                 ...                                       
2474                                            DESPAIR                                       
2475                                            DESPAIR                                       
2476                                            DESPAIR                                       
2477                                          NO_ANSWER                                       
2478                                            DESPAIR                                       

     Q6 | Any full-sized candy bar Q6 | Black Jacks Q6 | Bonkers (the candy)  \
1                              JOY              MEH                  DESPAIR   
2                        NO_ANSWER        NO_ANSWER                NO_ANSWER   
3                              JOY              MEH                      MEH   
4                              JOY          DESPAIR                      MEH   
5                              JOY        NO_ANSWER                NO_ANSWER   
...                            ...              ...                      ...   
2474                           MEH          DESPAIR                  DESPAIR   
2475                           JOY        NO_ANSWER                NO_ANSWER   
2476                           JOY          DESPAIR                      MEH   
2477                     NO_ANSWER        NO_ANSWER                NO_ANSWER   
2478                           JOY          DESPAIR                  DESPAIR   

     Q6 | Bonkers (the board game)  ... Q6 | Whatchamacallit Bars  \
1                          DESPAIR  ...                   DESPAIR   
2                        NO_ANSWER  ...                 NO_ANSWER   
3                          DESPAIR  ...                       JOY   
4                          DESPAIR  ...                       JOY   
5                        NO_ANSWER  ...                       JOY   
...                            ...  ...                       ...   
2474                           MEH  ...                   DESPAIR   
2475                     NO_ANSWER  ...                       JOY   
2476                           JOY  ...                       MEH   
2477                     NO_ANSWER  ...                 NO_ANSWER   
2478                       DESPAIR  ...                   DESPAIR   

     Q6 | White Bread Q6 | Whole Wheat anything Q6 | York Peppermint Patties  \
1             DESPAIR                   DESPAIR                      DESPAIR   
2           NO_ANSWER                

**Code Check:** As a check for the above code, make sure that there are no missing values left for the `Q6` to `Q12` columns.  

In [24]:
candy.loc[:, 'Q6 | 100 Grand Bar' : 'Q12: MEDIA [Yahoo]'].isna().sum()


Q6 | 100 Grand Bar                                                                        0
Q6 | Anonymous brown globs that come in black and orange wrappers\t(a.k.a. Mary Janes)    0
Q6 | Any full-sized candy bar                                                             0
Q6 | Black Jacks                                                                          0
Q6 | Bonkers (the candy)                                                                  0
                                                                                         ..
Q11: DAY                                                                                  0
Q12: MEDIA [Daily Dish]                                                                   0
Q12: MEDIA [Science]                                                                      0
Q12: MEDIA [ESPN]                                                                         0
Q12: MEDIA [Yahoo]                                                                        0
Length: 109, dtype: int64

Now, let's look at the very messy `Q4: COUNTRY` column and see what we can do about it. First, run the code below to look at the different unique values in the data.

In [25]:
# check unique values
candy['Q4: COUNTRY'].unique()

array(['USA ', 'USA', 'us', 'usa', nan, 'canada', 'Canada', 'Us', 'US',
       'Murica', 'United States', 'uk', 'United Kingdom', 'united states',
       'Usa', 'United States ', 'United staes',
       'United States of America', 'UAE', 'England', 'UK', 'canada ',
       'United states', 'u.s.a.', '35', 'france',
       'United States of America ', 'america', 'U.S.A.', 'finland',
       'unhinged states', 'Mexico', 'Canada ', 'united states of america',
       'US of A', 'The United States', 'North Carolina ', 'Unied States',
       'Netherlands', 'germany', 'Europe', 'U S', 'u.s.', 'U.K. ',
       'Costa Rica', 'The United States of America', 'unite states',
       'U.S.', '46', 'Australia', 'Greece', 'USA? Hard to tell anymore..',
       "'merica", '45', 'United State', '32', 'France', 'australia',
       'Can', 'Canae', 'Trumpistan', 'Ireland', 'United Sates', 'Korea',
       'California', 'Unites States', 'Japan', 'USa', 'South africa',
       'I pretend to be from Canada, but I am

**Code Check:** As a check for the Country column, check to see how many unique values are in the data.  You should have `115` different unique values for the `Q4: COUNTRY` column.  If you have less or more than this number, double check your work above.

In [26]:
# check the Q4: COUNTRY number of unique values
candy['Q4: COUNTRY'].nunique()

115

We want to clean up this data to only include four areas: USA, Canada, Europe (the continent, not necessarily the European Union), and Other.

There are different ways to do this, but I would suggest that you look at the way we handled the `property_type` column in the `vienna` data set and the code in the `amenities_to_columns()` function in the module notebook.  These might be a little harder than those examples but they should give you a good baseline approach.  

You could use `replace()` for this step, and it is fine if you ultimately decide to do this, but I would suggest that you come up with a solution similar to what was shown in the `vienna` data cleaning notebook.  This method would be much more robust if you had many more values in your data.

I suggest the following order for this section to make it easier:
- Fill in all missing values with `Other`
- Code Australia as `Other` (doing this step will help when trying to use `us` in the next step if you use string methods)
- Combine all USA entries together as `USA`
- Combine Canadian entries as `CA`
- Combine European entries as `EU`
- Everything else gets coded as `Other`

**Exercise11:** Fill the missing values in the `Q4: COUNTRY` column with `Other`.

In [27]:
candy.fillna({'Q4: COUNTRY' : 'Other'}, inplace= True)

**Code Check:** Double check that there are no missing values in the `Q4: COUNTRY` column.  Also, double check the unique values to make sure that "Other" was added.  This should mean that you now have `116` unique values for this column.

In [28]:
# check missing Q4 values
candy['Q4: COUNTRY'].isna().sum()


0

In [29]:
# check unique values
candy['Q4: COUNTRY'].unique()


array(['USA ', 'USA', 'us', 'usa', 'Other', 'canada', 'Canada', 'Us',
       'US', 'Murica', 'United States', 'uk', 'United Kingdom',
       'united states', 'Usa', 'United States ', 'United staes',
       'United States of America', 'UAE', 'England', 'UK', 'canada ',
       'United states', 'u.s.a.', '35', 'france',
       'United States of America ', 'america', 'U.S.A.', 'finland',
       'unhinged states', 'Mexico', 'Canada ', 'united states of america',
       'US of A', 'The United States', 'North Carolina ', 'Unied States',
       'Netherlands', 'germany', 'Europe', 'U S', 'u.s.', 'U.K. ',
       'Costa Rica', 'The United States of America', 'unite states',
       'U.S.', '46', 'Australia', 'Greece', 'USA? Hard to tell anymore..',
       "'merica", '45', 'United State', '32', 'France', 'australia',
       'Can', 'Canae', 'Trumpistan', 'Ireland', 'United Sates', 'Korea',
       'California', 'Unites States', 'Japan', 'USa', 'South africa',
       'I pretend to be from Canada, but 

**Exercise12:** Combine all Australia entries into `Other`.  Watch out for capitalization issues.  You should have `114` unique values after this step.

In [30]:
candy['Q4: COUNTRY'].replace({
    'Australia': 'Other', 'australia': 'Other'
}, inplace= True)

In [31]:
# check number of unique values
candy['Q4: COUNTRY'].unique()


array(['USA ', 'USA', 'us', 'usa', 'Other', 'canada', 'Canada', 'Us',
       'US', 'Murica', 'United States', 'uk', 'United Kingdom',
       'united states', 'Usa', 'United States ', 'United staes',
       'United States of America', 'UAE', 'England', 'UK', 'canada ',
       'United states', 'u.s.a.', '35', 'france',
       'United States of America ', 'america', 'U.S.A.', 'finland',
       'unhinged states', 'Mexico', 'Canada ', 'united states of america',
       'US of A', 'The United States', 'North Carolina ', 'Unied States',
       'Netherlands', 'germany', 'Europe', 'U S', 'u.s.', 'U.K. ',
       'Costa Rica', 'The United States of America', 'unite states',
       'U.S.', '46', 'Greece', 'USA? Hard to tell anymore..', "'merica",
       '45', 'United State', '32', 'France', 'Can', 'Canae', 'Trumpistan',
       'Ireland', 'United Sates', 'Korea', 'California', 'Unites States',
       'Japan', 'USa', 'South africa',
       'I pretend to be from Canada, but I am really from the Unite

**Exercise13:** Combine all United States entries together into `USA`.  These would include the following:
```
'USA ', 'USA', 'us', 'usa', 'Us', 'US', 'Murica', 'United States', 'united states', 'Usa', 'United States ', 'United staes', 'United States of America', 'United states', 'u.s.a.', 'United States of America ', 'america', 'U.S.A.', 'unhinged states', 'united states of america', 'US of A', 'The United States', 'North Carolina ', 'Unied States', 'U S', 'u.s.', 'The United States of America', 'unite states','U.S.', 'USA? Hard to tell anymore..', "'merica", 'United State', 'United Sates', 'California', 'Unites States', 'USa', 'I pretend to be from Canada, but I am really from the United States.', 'Usa ', 'United Stated', 'New Jersey', 'United ststes', 'America', 'United Statss', 'murrika', 'USA! USA! USA!', 'USAA', 'united States ', 'N. America', 'USSA', 'U.S. ', 'u s a', 'United Statea', 'united ststes', 'USA USA USA!!!!'
```

**Careful:** Some students try to use regex for this step with a string method.  Be careful as there is one response with the term "subscribe" in it.  If you create your regex incorrectly, your use of "u.s." could mistakenly pick up this record because of the "ubs" contained in it and code it as "USA".

In [32]:
candy['Q4: COUNTRY'] = candy['Q4: COUNTRY'].str.lower()

candy['Q4: COUNTRY'].replace({
    'usa ': 'USA', 'usa': 'USA', 'us': 'USA', 'murica': 'USA', 'united states': 'USA',
    'united states ': 'USA', 'united staes': 'USA', 'united states of america': 'USA',
    'america': 'USA', 'u.s.a.': 'USA', 'unhinged states': 'USA', 'us of a': 'USA',
    'the united states': 'USA', 'north carolina ': 'USA', 'unied states': 'USA', 'u s': 'USA',
    'u.s.': 'USA', 'usa? hard to tell anymore..': 'USA', "'merica": 'USA', 'unite states': 'USA',
    'california': 'USA', 'i pretend to be from canada, but i am really from the united states.': 'USA',
    'new jersey': 'USA', 'united ststes': 'USA', 'united statss': 'USA', 'murrika': 'USA',
    'usa! usa! usa!': 'USA', 'usaa': 'USA', 'n. america': 'USA', 'ussa': 'USA', 'u.s. ': 'USA',
    'u s a': 'USA', 'united statea': 'USA', 'united ststes': 'USA', 'usa usa usa!!!!': 'USA',
    'united states of america': 'USA', 'the united states of america': 'USA', 'united state': 'USA',
    'united sates': 'USA', 'unites states': 'USA', 'united stated': 'USA', 'united states of america ':'USA'
}, inplace=True)



**Code Check:** You should be merging the above values together into 1 (`USA`) and be left with 61 unique values after this step (including the `USA` value).

In [33]:
# check unique values
candy['Q4: COUNTRY'].unique()



array(['USA', 'other', 'canada', 'uk', 'united kingdom', 'uae', 'england',
       'canada ', '35', 'france', 'finland', 'mexico', 'netherlands',
       'germany', 'europe', 'u.k. ', 'costa rica', '46', 'greece', '45',
       '32', 'can', 'canae', 'trumpistan', 'ireland', 'korea', 'japan',
       'south africa', 'canada`', 'scotland', 'uk ', 'denmark', 'france ',
       'switzerland', 'ud', 'scotland ', 'south korea', 'indonesia',
       'the netherlands', 'endland', 'soviet canuckistan', 'singapore',
       'china', 'taiwan', 'ireland ', 'hong kong', 'spain', 'sweden',
       'narnia', 'subscribe to dm4uz3 on youtube', "i don't know anymore",
       'fear and loathing'], dtype=object)

**Exercise14:** Combine the Canadian entries (both upper and lower case) and label them as `CA`. Be careful as there are extra spaces, characters, and misspellings (Can, Canae).

These values include:
```
'canada', 'Canada', 'canada ', 'Canada ', 'Can', 'Canae', 'Canada`', 'CANADA'
```

In [34]:
candy['Q4: COUNTRY'].replace({
    'canada': 'CA', 'Canada': 'CA', 'canada ': 'CA', 'Can': 'CA',
    'Canae': 'CA', 'Canada`': 'CA', 'CANADA': 'CA', 'can': 'CA',
    'canae': 'CA','canada`': 'CA', 'ca': 'CA'
}, inplace= True)

**Code Check:** You should be merging 8 values together into 1 (`CA`) and be left with 54 unique values after this step (including the `CA` value).

In [35]:
# check unique values
candy['Q4: COUNTRY'].unique()

array(['USA', 'other', 'CA', 'uk', 'united kingdom', 'uae', 'england',
       '35', 'france', 'finland', 'mexico', 'netherlands', 'germany',
       'europe', 'u.k. ', 'costa rica', '46', 'greece', '45', '32',
       'trumpistan', 'ireland', 'korea', 'japan', 'south africa',
       'scotland', 'uk ', 'denmark', 'france ', 'switzerland', 'ud',
       'scotland ', 'south korea', 'indonesia', 'the netherlands',
       'endland', 'soviet canuckistan', 'singapore', 'china', 'taiwan',
       'ireland ', 'hong kong', 'spain', 'sweden', 'narnia',
       'subscribe to dm4uz3 on youtube', "i don't know anymore",
       'fear and loathing'], dtype=object)

**Exercise15:** Combine the European entries and label them as `EU`. Again, we are looking at the continent of Europe and not necessarily the countries that are a part of the European Union.  

These values include:
```
'uk', 'United Kingdom', 'England', 'UK', 'france', 'finland', 'Netherlands', 'germany', 'Europe', 'U.K. ', 'Greece', 'France', 'Ireland', 'Uk', 'Germany', 'Scotland', 'UK ', 'Denmark', 'France ', 'Switzerland', 'Scotland ', 'The Netherlands', 'Ireland ', 'spain', 'Sweden', 'United kingdom'
```

In [36]:
candy['Q4: COUNTRY'].replace({
    'uk': 'EU', 'united kingdom': 'EU', 'england': 'EU', 'UK': 'EU', 'france': 'EU',
    'finland': 'EU', 'netherlands': 'EU', 'germany': 'EU', 'europe': 'EU', 'u.k.': 'EU',
    'greece': 'EU', 'ireland': 'EU', 'scotland': 'EU', 'denmark': 'EU', 'switzerland': 'EU',
    'the netherlands': 'EU', 'spain': 'EU', 'sweden': 'EU', 'united kingdom': 'EU', 'u.k. ': 'EU',
    'ireland ': 'EU',  'france ': 'EU', 'scotland ': 'EU', 'uk ': 'EU', 'endland': 'EU', 'eu': 'EU'
}, inplace= True)

**Code Check:** You should be merging 26 entries together and be left with 29 unique values after this step (including the `EU` value).

In [37]:
# check unique values
candy['Q4: COUNTRY'].unique()

array(['USA', 'other', 'CA', 'EU', 'uae', '35', 'mexico', 'costa rica',
       '46', '45', '32', 'trumpistan', 'korea', 'japan', 'south africa',
       'ud', 'south korea', 'indonesia', 'soviet canuckistan',
       'singapore', 'china', 'taiwan', 'hong kong', 'narnia',
       'subscribe to dm4uz3 on youtube', "i don't know anymore",
       'fear and loathing'], dtype=object)

**Exercise16:** Finally, combine the other entries and label them as `Other`.

In [38]:
candy['Q4: COUNTRY'].replace({
    'uae': 'Other', '35': 'Other', 'mexico': 'Other', 'costa rica': 'Other',
    '46': 'Other', '45': 'Other', '32': 'Other',
    'trumpistan': 'Other', 'korea': 'Other', 'japan': 'Other',
    'south africa': 'Other', 'ud': 'Other', 'south korea': 'Other',
    'indonesia': 'Other', 'soviet canuckistan': 'Other', 'singapore': 'Other',
    'china': 'Other', 'taiwan': 'Other', 'hong kong': 'Other',
    'narnia': 'Other', 'subscribe to dm4uz3 on youtube': 'Other',
    "i don't know anymore": 'Other', 'fear and loathing': 'Other', 'other': 'Other'
}, inplace= True)

**Code Check:** Double check that you only have four unique values in the `Q4: COUNTRY` column: `USA`, `Other`, `CA`, and `EU`

In [39]:
# check values
candy['Q4: COUNTRY'].unique()

array(['USA', 'Other', 'CA', 'EU'], dtype=object)

**Q5:** To double check that everything was coded correctly, save the value counts of the `Q4: COUNTRY` column as `Q5`.  You can check this once you run your CodeGrade check.

In [40]:
Q5 = candy['Q4: COUNTRY'].value_counts()

We now want to look at the `Q3: AGE` column. Let's look at all the unique values.

In [41]:
# check unique age values
candy['Q3: AGE'].unique()

array(['44', '49', '40', '23', nan, '53', '33', '43', '56', '64', '37',
       '48', '54', '36', '45', '25', '34', '35', '38', '58', '50', '47',
       '16', '52', '63', '65', '41', '27', '31', '59', '61', '46', '42',
       '62', '29', '39', '32', '28', '69', '67', '30', '22', '51', '70',
       '24', '19', 'Old enough', '57', '60', '66', '12', 'Many', '55',
       '72', '?', '21', '11', 'no', '9', '68', '20', '6', '10', '71',
       '13', '26', '45-55', '7', '39.4', '74', '18', 'older than dirt',
       '17', '15', '8', '75', '5u', 'Enough', 'Over 50', '90', '76',
       'sixty-nine', 'ancient', '77', 'OLD', 'old', '73', '70 1/2', '14',
       'MY NAME JEFF', '4', '59 on the day after Halloween', 'old enough',
       'your mom', 'I can remember when Java was a cool new language',
       '60+'], dtype=object)

Again, this is a pretty messy column of data. This is a good example of why those that create online surveys shouldn't allow the individual to just put any value into the field. But it is now our job to clean this up.

**Exercise17:** Your task is to put these values into the following categorical bins: `unknown`, `17 and under`, `18-25`, `26-35`, `36-45`, `46-55`, and `56+`.

- The category labels should exactly match the above.
- Missing values should be replaced with the `unknown` category
- To make things easier and avoid ambiguity, let's say that any value with text, even if we could determine the age, will be binned with the `unknown` category. For example: `sixty-nine` should be coded as `unknown`, `45-55` should be coded as `unknown`, `59 on the day after Halloween` should be coded as `unknown`, etc.
- Ensure that the category labels are unordered but reorder the categories so that 'unknown' is listed in the first position. This is not really needed but will help us grade your assignment. The categories should be listed as follows: `Index(['unknown', '17 and under', '18-25', '26-35', '36-45', '46-55', '56+'], dtype='object')`

First, we will replace any non-numeric value (those with text as mentioned above) with a missing value.  This will allow you to turn the other values into floats so that you can bin them. Just don't forget to code the missing values as `unknown` when you are done.  To replace the non-numeric values, run the following code:

In [42]:
### DO NOT CHANGE THIS CODE - RUN AS IS ###
### This code replaces non-numeric values with a missing value to make your next step easier ###

# create True/False index
age_index = candy['Q3: AGE'].str.isnumeric()

# for the index, fill missing values with False
age_index = age_index.fillna(False)

# select Age column for only those False values from index and code as missing
candy.loc[~age_index, 'Q3: AGE'] = np.nan

Now, you can finish exercise 17 to bin the ages and write the code below:

In [50]:
bins = [0, 17, 25, 35, 45, 55, float('inf')]
labels = ['17 and under', '18-25', '26-35', '36-45', '46-55', '56+']

candy['Q3: AGE'] = pd.cut(candy['Q3: AGE'], bins=bins, labels=labels, right = False)

candy['Q3: AGE'] = candy['Q3: AGE'].cat.add_categories('unknown').fillna('unknown')

print(candy['Q3: AGE'].cat.categories)

Index(['17 and under', '18-25', '26-35', '36-45', '46-55', '56+', 'unknown'], dtype='object')


**Exercise18:** Double check yourself by checking the categories for the `Q3: AGE` column. It should output: `Index(['unknown', '17 and under', '18-25', '26-35', '36-45', '46-55', '56+'], dtype='object')`

In [51]:
# double check categories
candy['Q3: AGE'].cat.categories

Index(['17 and under', '18-25', '26-35', '36-45', '46-55', '56+', 'unknown'], dtype='object')

**Code Check:** To double check your above binning worked correctly, your value counts (sorted by the index) should be as follows:

```
unknown: 60
17 and under: 49
18-25: 85
26-35: 520
36-45: 768
46-55: 525
56+: 298
```

In [52]:
candy['Q3: AGE'].value_counts(sort=False)

Q3: AGE
17 and under     46
18-25            72
26-35           464
36-45           770
46-55           563
56+             330
unknown          60
Name: count, dtype: int64

You can also double check some of your work up to this point by making sure that there are no missing values in the data set anymore.

**Code Check:** Check to see if there are any missing values in the data set. Your output should show `0`.

In [53]:
candy.isnull().sum()

Q1: GOING OUT?             0
Q2: GENDER                 0
Q3: AGE                    0
Q4: COUNTRY                0
Q6 | 100 Grand Bar         0
                          ..
Q11: DAY                   0
Q12: MEDIA [Daily Dish]    0
Q12: MEDIA [Science]       0
Q12: MEDIA [ESPN]          0
Q12: MEDIA [Yahoo]         0
Length: 113, dtype: int64

**Exercise19:** Before you move on to the next section, reset the index for `candy` ensuring that it goes from 0 to n-1.  

In [54]:
candy.reset_index(drop= True, inplace= True)

## Feature Engineering

Feature engineering is the process of transforming raw data into features that better represent the underlying problem to the predictive models.  In this section, we will create a new column called "net_feelies" (calculated by the authors as the total joy count minus the total despair count).

First, let's narrow down our data to make working with it easier.

**Exercise20:** Select only the Q6 candy columns (`Q6 | 100 Grand Bar` through `Q6 | York Peppermint Patties`) in the data set and save this as a new DataFrame called `candy_reduced`.

In [58]:
candy_reduced = candy.loc[:, 'Q6 | 100 Grand Bar' : 'Q6 | York Peppermint Patties']

Next, we will create two Series, one with JOY counts and one with DESPAIR counts to add to our `candy_reduced` data.

**Exercise21:** Create a Series called `joy_count` that lists total counts for JOY for each column, making sure to keep it in the same order as the columns in the `candy_reduced` DataFrame. Hint: A simple way to do this is to filter the entire DataFrame for any `JOY` values and then use `count()`. See this [stackoverflow question](https://stackoverflow.com/questions/63103090/how-do-i-count-specific-values-across-multiple-columns-in-pandas) and answers.

In [59]:
joy_count = candy_reduced.apply(lambda x: (x == 'JOY').sum())

**Exercise22:** Same as above except you will create a Series called `despair_count` that lists the total counts for DESPAIR for each column.

In [60]:
despair_count = candy_reduced.apply(lambda x: (x == 'DESPAIR').sum())

**Exercise23:** Take the transpose of the `candy_reduced` DataFrame and save this transposed data as `candy_reduced_transpose`.

In [61]:
candy_reduced_transpose = candy_reduced.transpose()

**Exercise24:** Add a new column called "joy_count" using the `joy_count` Series above and a new column called 'despair_count" using the `despair_count` Series above to the `candy_reduced_transpose` DataFrame.

In [62]:
candy_reduced_transpose['joy_count'] = joy_count

candy_reduced_transpose['despair_count'] =despair_count

**Exercise25:** Add a new column to the `candy_reduced_transpose` DataFrame called "net_feelies" that takes the `joy_count` column and subtracts the `despair_count` column.

In [63]:
candy_reduced_transpose['net_feelies'] = candy_reduced_transpose['joy_count'] - candy_reduced_transpose['despair_count']

**Exercise26:** Select only the `joy_count`, `despair_count`, and `net_feelies` columns from the `candy_reduced_transpose` DataFrame. Sort this DataFrame in descending order by `net_feelies` and save this as `candy_net_sorted`.

In [64]:
candy_net_sorted = candy_reduced_transpose[['joy_count', 'despair_count', 'net_feelies']]

candy_net_sorted = candy_net_sorted.sort_values(by= 'net_feelies', ascending= False)

## Encoding

We now want to get the `candy` DataFrame ready to run a machine learning algorthim to determine if we could predict a person's gender based on what candy they prefer.

We will now turn all of our values into numeric values.  There are easier ways to do this with sklearn, but we will use Pandas to perform these exercises for further practice.

**Exercise27:** For grading purposes, we want to leave the `candy` DataFrame as is. Make a copy of the `candy` DataFrame and save this new DataFrame as `candy_encode`.

In [65]:
candy_encode = candy.copy()

**Exercise28:** For the `candy_encode` DataFrame, replace any `Female` values with `0` and any `Male` values with `1`.

In [66]:
candy_encode['Q2: GENDER'] = candy_encode['Q2: GENDER'].replace({'Female': 0, 'Male': 1})

**Exercise29:** Again, you will learn more about this later, but we need to separate the column that we want to predict (called the response) and the columns that we will use to make the predictions (called the features).  **For both of the items below, make sure that the index is reset and goes from 0 to n-1.**

- Select only the `Q2: GENDER` column from `candy_encode` and save this as `candy_response`.  **Note: This should be a Series.**
- Drop the following columns from the `candy_encode` DataFrame: `Q2: GENDER`,`Q1: GOING OUT?`,`Q3: AGE`,`Q4: COUNTRY`,`Q10: DRESS`,`Q11: DAY`, `Q12: MEDIA [Daily Dish]`,`Q12: MEDIA [Science]`,`Q12: MEDIA [ESPN]`,`Q12: MEDIA [Yahoo]`.  Save the remaining columns as `candy_features`.

In [67]:
candy_response = candy_encode['Q2: GENDER']

enconde_columns_removed= [
    'Q2: GENDER', 'Q1: GOING OUT?', 'Q3: AGE',
    'Q4: COUNTRY', 'Q10: DRESS', 'Q11: DAY',
    'Q12: MEDIA [Daily Dish]', 'Q12: MEDIA [Science]',
    'Q12: MEDIA [ESPN]', 'Q12: MEDIA [Yahoo]'
]

candy_features = candy_encode.drop(columns=enconde_columns_removed)

**Exercise30:** Use Panda's `get_dummies()` to encode the `candy_features` data, making sure to set `drop_first=True` and `dtype=int`. Save this as `candy_features_encoded`.

In [70]:
candy_features_encoded = pd.get_dummies(candy_features, drop_first= True, dtype= int)

**Code Check:** Make sure that the `candy`, `candy_features`, `candy_response` and `candy_features_encoded` have an index that goes from 0 to n-1 or your final CodeGrade tests will not pass.

## Final Analysis

THe data is now cleaned and prepared to be passed to a machine learning model.  

I created models using Random Forest, Logistic Regression, and XGBoost algorithms, and they all returned around 70% accuracy rates. However, the other accuracy metrics didn't look as good. Given the metrics that were calculated, I would say that based only on this data, using candy preference is not that great of an indicator of someone's gender.
